### PR TITLE
Moves standard URNs into .proto files

### PR DESCRIPTION
**Pull request by jkff**
Centralizes the location of standard URNs - now they are encoded in .proto files, implemented as options on enum values. This way they are located only in one place, and references are compile-time checked (where available), rather than hard-coded in many places. `common_urns.md` was a step in this direction, but it only validated that an URN is valid, but did not provide a named language entity for each one.

Fixes a bunch of code that referred to them by old URNs.

This removes the necessity for UrnUtils and for the `resources` directory.

### CODE REVIEW

**Comment by jkff** [comment]
CC: @herohde

**Comment by jkff** [comment]
OK, fixed the build dependencies in build.gradle too I think.

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 1]
```diff
@@ -175,6 +176,72 @@ message PTransform {
   DisplayData display_data = 6;
 }
 
+message StandardPTransformUrns {
```
Comment (kennknowles) [inline model/pipeline/src/main/proto/beam_runner_api.proto]: Separate primitives, deprecated primitives, standardized composites.

Primitives are: ParDo, Flatten, GroupByKey, Impulse, AssignWindows

Deprecated primitives are: Read, CreateView

There's some mystery surrounding the status of GBK / CoGBK

The rest are just composite that have well known URNs. They should live as annotations closer to their payloads, but not alongside the primitives.

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 2]
```diff
@@ -175,6 +176,72 @@ message PTransform {
...
+    RESHUFFLE = 13 [(standard_urn) = "beam:transform:reshuffle:v1"];
+
+    // Less well-known. Payload: WriteFilesPayload.
+    WRITE_FILES = 14 [(standard_urn) = "beam:transform:write_files:0.1"];
+
+    // Deprecated: runners should move away from translating `CreatePCollectionView` and treat this as
+    // part of the translation for a `ParDo` side input.
+    CREATE_VIEW = 15 [(standard_urn) = "beam:transform:create_view:v1"];
+
+    MAP_WINDOWS = 16 [(standard_urn) = "beam:transform:map_windows:v1"];
```
Comment (kennknowles) [inline model/pipeline/src/main/proto/beam_runner_api.proto]: Don't even know what this is

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 3]
```diff
@@ -175,6 +176,72 @@ message PTransform {
...
+
+    // Represents the Combine.groupedValues() operation.
+    // If this is produced by an SDK, it is assumed that the SDK understands pgbkcv,
+    // merge_accumulators and extract_outputs.
+    // Payload: CombinePayload
+    COMBINE_GROUPED_VALUES = 9 [(standard_urn) = "beam:transform:combine_grouped_values:v1"];
+
+    COMBINE_PGBKCV = 10 [(standard_urn) = "beam:transform:combine_pgbkcv:v1"];
+    COMBINE_MERGE_ACCUMULATORS = 11 [(standard_urn) = "beam:transform:combine_merge_accumulators:v1"];
+    COMBINE_EXTRACT_OUTPUTS = 12 [(standard_urn) = "beam:transform:combine_extract_outputs:v1"];
```
Comment (kennknowles) [inline model/pipeline/src/main/proto/beam_runner_api.proto]: Separate fn api-only URNs from pipeline-level URNs

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 4]
```diff
@@ -403,6 +470,33 @@ message Coder {
   repeated string component_coder_ids = 2;
 }
 
+message StandardCoderUrns {
```
Comment (kennknowles) [inline model/pipeline/src/main/proto/beam_runner_api.proto]: I believe these are all actually required, but check to be sure.

Missing TIMESTAMP (I think durations are not primitives)

File: model/pipeline/src/main/proto/standard_window_fns.proto [hunk 5]
```diff
@@ -29,26 +29,41 @@ option go_package = "pipeline_v1";
...
 option java_outer_classname = "StandardWindowFns";
 
+import "beam_runner_api.proto";
 import "google/protobuf/duration.proto";
 import "google/protobuf/timestamp.proto";
 
-// beam:windowfn:global_windows:v0.1
-// empty payload
+message StandardWindowFnUrns {
+  enum Enum {
```
Comment (kennknowles) [inline model/pipeline/src/main/proto/standard_window_fns.proto]: This is an interesting example. No WindowFn is required as a primitive. They are all just "well-known". There is no particular reason they should be in the same list, except the SDK *today* ships with them.

What about putting these in an enum inside their payload?

```
message FixedWindows {
   enum Urn {
     FIXED_WINDOWS = 0 [(beam_urn) = "beam:windowfn:fixed_windows:v0.1"]
   }
   google.protobuf.Duration size = 1;
   google.protobuf.Timestamp offset = 2;	   google.protobuf.Timestamp offset = 2;
}
```

(the `Payload` suffix has always been extraneous; I regret introducing it)

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 6]
```diff
@@ -748,6 +842,22 @@ message SdkFunctionSpec {
...
+  // An extension to be used for specifying the standard URN of various
+  // pipeline entities, e.g. transforms, functions, coders etc.
+  // Code should refer to the URNs of those entities by extracting
+  // it from the (standard_urn) extension, rather than by hard-coding
+  // the URN.
+  //
+  // The recommended pattern for declaring it is (exemplified by coders):
+  //
+  // enum StandardCoderUrns {
+  //   BYTES = 0 [(standard_urn) = "beam:coder:bytes:v1"];
```
Comment (kennknowles) [inline model/pipeline/src/main/proto/beam_runner_api.proto]: I think `beam_urn` might be better than `standard_urn`

File: model/pipeline/src/main/proto/beam_runner_api.proto (hunk 3 above)
Comment (robertwb) [reply model/pipeline/src/main/proto/beam_runner_api.proto]: Disagree here, much better to document all the combine URNs together than some in one place and some in another.

File: model/pipeline/src/main/proto/beam_runner_api.proto (hunk 1 above)
Comment (robertwb) [reply model/pipeline/src/main/proto/beam_runner_api.proto]: +1

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 7]
```diff
@@ -175,6 +176,72 @@ message PTransform {
...
+    COMBINE_MERGE_ACCUMULATORS = 11 [(standard_urn) = "beam:transform:combine_merge_accumulators:v1"];
+    COMBINE_EXTRACT_OUTPUTS = 12 [(standard_urn) = "beam:transform:combine_extract_outputs:v1"];
+
+    // Represents the Reshuffle operation.
+    RESHUFFLE = 13 [(standard_urn) = "beam:transform:reshuffle:v1"];
+
+    // Less well-known. Payload: WriteFilesPayload.
+    WRITE_FILES = 14 [(standard_urn) = "beam:transform:write_files:0.1"];
+
+    // Deprecated: runners should move away from translating `CreatePCollectionView` and treat this as
```
Comment (robertwb) [inline model/pipeline/src/main/proto/beam_runner_api.proto]: Let's not put known-deprecated URNs here, but rather in a separate file.

File: model/pipeline/src/main/proto/standard_window_fns.proto (hunk 5 above)
Comment (robertwb) [reply model/pipeline/src/main/proto/standard_window_fns.proto]: We may require GlobalWindows (e.g. when we need to pass elements and don't care about the windowing). 

Putting these URNs as part of the payload is an interesting idea.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/WindowingStrategyTranslation.java [hunk 8]
```diff
@@ -365,35 +358,33 @@ public static SdkFunctionSpec toProto(WindowFn<?, ?> windowFn, SdkComponents com
...
-              Duration.millis(Durations.toMillis(sessionParams.getGapSize())));
-        case SERIALIZED_JAVA_WINDOWFN_URN:
-          return (WindowFn<?, ?>)
-              SerializableUtils.deserializeFromByteArray(
-                  windowFnSpec.getSpec().getPayload().toByteArray(), "WindowFn");
-        default:
-          throw new IllegalArgumentException(
-              "Unknown or unsupported WindowFn: " + windowFnSpec.getSpec().getUrn());
+      String s = windowFnSpec.getSpec().getUrn();
+      if (s.equals(getStandardUrn(StandardWindowFnUrns.Enum.GLOBAL_WINDOWS))) {
```
Comment (robertwb) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/WindowingStrategyTranslation.java]: We could remove the case statements, but inlining getStandardUrn(...) is quite verbose here.

File: sdks/python/setup.py [hunk 9]
```diff
@@ -148,35 +148,6 @@ def run(self):
...
 
 
-def generate_common_urns():
-  src = os.path.join(
-      os.path.dirname(__file__),
-      '../../'
-      'model/pipeline/src/main/resources/org/apache/beam/model/common_urns.md')
-  out = os.path.join(
-    os.path.dirname(__file__),
-    'apache_beam/portability/common_urns.py')
```
Comment (robertwb) [inline sdks/python/setup.py]: If we're not auto-generating this, we need to check it in (or something else that extracts all the names from the annotations).

File: model/pipeline/src/main/proto/beam_runner_api.proto (hunk 3 above)
Comment (kennknowles) [reply model/pipeline/src/main/proto/beam_runner_api.proto]: I can accept this. They should live perhaps in an enum inside or alongside the `Combine` payload.

As you know, I broadly believe there is value is being clear about what is a PCollection(s)-to-PCollection(s) transformation versus what is an instruction that we can tell an SDK harness to apply to a bundle. Losing that distinction has already led to proposals that things that used to be automatic - like combiner lifting - being pushed back into the submitted pipeline.

So my only issue is I can't really tell from this PR if we are taking the position that these are all valid PCollection(s)-to-PCollection(s) transformations that just happen to only be used as instructions, or if we are just not being clear about the distinction.

File: sdks/python/setup.py (hunk 9 above)
Comment (jkff) [reply sdks/python/setup.py]: Oh snap. I *did* check it in, but it is ignored in .gitignore (forgot to remove this) and cleaned by "mvn clean" (remembered to remove from there when it was too late) :-/ Guess I'll have to redo it.

Comment (robertwb) [reply sdks/python/setup.py]: We should not just check this in, that would be a regression. The key point is that there's a common source of truth (namely the protos).

Comment (robertwb) [reply sdks/python/setup.py]: (I mean, checking in the auto-generated file. I assume what you did is rewrite this file to extract the proto annotations.)

**Comment by jkff** [comment]
(I'll resume working on this next week - currently busy preparing the Strata talk)

**Comment by jkff** [comment]
OK, I refactored this a little: categorized the entities into different enums in the same message where appropriate (reads better, and allows to potentially have more properties specified as an option than just URN), and re-created the lost common_urns.py. PTAL.

**Comment by jkff** [comment]
Ping.

**Comment by jkff** [comment]
Rebased and revised, PTAL :)

**Comment by jkff** [comment]
Run Gradle Java PreCommit

**Comment by jkff** [comment]
retest this please

**Comment by jkff** [comment]
retest this please

**Comment by jkff** [comment]
Thanks! Oh, realized I forgot to refactor the Go code the same way. I'll file another JIRA for that alongside those that you recommended.

**Pull request merged**